shared-module/audiofilters/Filter.c

```diff
@@ -0,0 +1,346 @@
+                // Track length in terms of words.
+                self->sample_buffer_length /= (self->bits_per_sample / 8);
+                self->more_data = result == GET_BUFFER_MORE_DATA;
+            }
+        }
+
+        // If we have a sample, filter it
+        if (self->sample != NULL) {
+            // Determine how many bytes we can process to our buffer, the less of the sample we have left and our buffer remaining
+            uint32_t n = MIN(self->sample_buffer_length, length);
+
+            int16_t *sample_src = (int16_t *)self->sample_remaining_buffer; // for 16-bit samples
+            int8_t *sample_hsrc = (int8_t *)self->sample_remaining_buffer; // for 8-bit samples
+
+            if (mix <= 0.01 || self->filter_obj == mp_const_none) { // if mix is zero pure sample only or no biquad filter object is provided
+                for (uint32_t i = 0; i < n; i++) {
+                    if (MP_LIKELY(self->bits_per_sample == 16)) {
+                        word_buffer[i] = sample_src[i];
+                    } else {
+                        hword_buffer[i] = sample_hsrc[i];
+                    }
+                }
+            } else {
+                uint32_t i = 0;
+                while (i < n) {
+                    uint32_t n_samples = MIN(SYNTHIO_MAX_DUR, n - i);
+
+                    // Fill filter buffer with samples
+                    for (uint32_t j = 0; j < n_samples; j++) {
+                        if (MP_LIKELY(self->bits_per_sample == 16)) {
+                            self->filter_buffer[0][j] = sample_src[i + j];
+                        } else {
+                            if (self->samples_signed) {
+                                self->filter_buffer[0][j] = sample_hsrc[i + j];
+                            } else {
+                                // Be careful here changing from an 8 bit unsigned to signed into a 32-bit signed
+                                self->filter_buffer[0][j] = (int8_t)(((uint8_t)sample_hsrc[i + j]) ^ 0x80);
+                            }
+                        }
+                    }
+
+                    // Copy original signal for mixing back in later
+                    memcpy(self->filter_buffer[1], self->filter_buffer[0], n_samples);
+
+                    // Process biquad filter
+                    synthio_biquad_filter_samples(&self->filter_state, self->filter_buffer[0], n_samples);
+
+                    // Mix processed signal with original sample and transfer to output buffer
+                    for (uint32_t j = 0; j < n_samples; j++) {
+                        int32_t word = (self->filter_buffer[1][j] * (1.0 - mix)) + (self->filter_buffer[0][j] * mix);
+                        if (MP_LIKELY(self->bits_per_sample == 16)) {
+                            word_buffer[i + j] = mix_down_sample(word);
```

Comment:
The `mix_down_sample` was used if the 32-bit value could be over the 16-bit size, but for a filter and with the mix will this ever be the case? I don't think a filter will ever make a single value larger then it started, and neither will the mix calculation.

Reply:
Filter resonance can cause spikes in the output (ie: self-oscillation). I'm not against taking out this calculation, but it may reduce unexpected peaks in some situations. However, removing it would certainly simplify the module.

Reply:
After some quick testing with `mix_down_sample` removed, I think it is best to keep it as is. Without it, the output experiences significant clipping within the example above (tested with a resonance of 2.0).

Reply:
Makes sense, just wanted to ensure it was required. Since we use this now in 3 places, I am thinking we need to look at making a generic version that can be called from everywhere. For me, doesn't have to be right now, but next time one of us updates a filter may be worth looking at.